feature: 라우팅 및  HomePage -> QuizPage 페이지이동, 레이아웃 , 컴포넌트 생성 , Header(퀴즈 이동 버튼 기능 구현)

nemo0824 opened the pull request:
feature: 
Header 컴포넌트 퀴즈 이동 버튼 기능 구현
라우터 설정 및 Home → QuizPage 이동
컴포넌트 생성(Header, Footer, Layout, Quiz, QuizEditor, QuizVisualizaiotn, sidebar, Home, QuizPage) 임시데이터 생성 (data.ts)

nemo0824 committed:
feature:
라우트팅 설정
Header 컴포넌트 퀴즈 이동 버튼 기능 구현
라우터 설정 및 Home  → QuizPage 이동
컴포넌트 생성(Header, Footer, Layout, Quiz, QuizEditor, QuizVisualizaiotn, sidebar, Home, QuizPage)
임시데이터 생성 (data.ts)

Jaehoo-dev commented on src/App.tsx:
```diff
@@ -1,13 +1,21 @@
+import { Routes, Route } from 'react-router-dom';
+import Layout from './components/Layout';
+import QuizPage from './page/QuizPage';
+import Home from './page/Home';
```
```suggestion
import Home from './page/Home';

```

new-deni replied:
```suggestion
import Home from './page/Home';

```

Jaehoo-dev commented on src/components/Footer.tsx:
```diff
@@ -0,0 +1,7 @@
+import React from 'react';
```
요것도 필요없지 않은가요?

Jaehoo-dev commented on src/components/Header.tsx:
```diff
@@ -0,0 +1,100 @@
+  const navigate = useNavigate();
+  const { quizId } = useParams();
+  const [currentQuizId, setCurrentQuizId] = useState<number>(
+    parseInt(quizId || '1', 10)
```
스트링 1을 기본값으로 넣는 건 어색해 보여요.
quizId가 항상 있다고 보장해야 한다면 검증을 해서 에러를 터트리고, 없을 수도 있고 없을 때 기본값을 넣는 흐름이라면 url 파람을 실제로 바꾸는 게 좋아 보여요.

Jaehoo-dev commented on src/components/Header.tsx:
```diff
@@ -0,0 +1,100 @@
+const Header = () => {
+  const navigate = useNavigate();
+  const { quizId } = useParams();
+  const [currentQuizId, setCurrentQuizId] = useState<number>(
+    parseInt(quizId || '1', 10)
+  );
+  const [loading, setLoading] = useState(false);
```
currentQuizId는 number라고 명시했는데 loading은 타입을 명시하지 않았군요? 타입스크립트가 알아서 추론하라는 방법도 좋고 명시하는 방법도 좋은데 차이를 알고 통일도 하는 게 좋아요~

new-deni replied:
currentQuizId는 number라고 명시했는데 loading은 타입을 명시하지 않았군요? 타입스크립트가 알아서 추론하라는 방법도 좋고 명시하는 방법도 좋은데 차이를 알고 통일도 하는 게 좋아요~

new-deni commented on src/App.tsx:
```diff
@@ -1,13 +1,21 @@
+import Home from './page/Home';
 function App() {
   return (
     <>
```
요거 필요한가요?

new-deni commented on src/components/Header.tsx:
```diff
@@ -0,0 +1,100 @@
+  );
+  const [loading, setLoading] = useState(false);
+  const [error, setError] = useState<string | null>(null);
+  const [currentQuiz, setCurrentQuiz] = useState<QuizType | null>(null);
```
currentQuizId와 currentQuiz를 모두 상태로 관리해야 하나요? 관리하는 상태가 많아질수록 복잡도가 높아지기 때문에 상태 개수는 적을수록 좋습니다.

new-deni commented on src/components/Header.tsx:
```diff
@@ -0,0 +1,100 @@
+  };
+
+  useEffect(() => {
+    setLoading(true);
+    fetchQuizData(currentQuizId);
+    setLoading(false);
```
이거 loading 상태 바꾸는 게 의도한 대로 동작하나요? 동작하는 것처럼 보여도 보장할 수 없을 거예요. 더해서 퀴즈 정보를 불러오는 일은 나중에 비동기로 동작할 텐데 그렇다면 목업으로서 역할을 못하고 있네요~

new-deni commented on src/components/Header.tsx:
```diff
@@ -0,0 +1,100 @@
+  }, [currentQuizId]);
+
+  const handlePrevious = () => {
+    if (currentQuiz && currentQuizId > 1) {
```
1이라는 매직넘버가 반복적으로 나오고 코드에서 흩뿌려져 있네요. 개선해보세요~

new-deni commented on src/components/Header.tsx:
```diff
@@ -0,0 +1,100 @@
+        <p>Javascript Playground</p>
+      </div>
+      <div className="absolute left-1/2 transform -translate-x-1/2 flex items-center gap-3">
+        {location.pathname === '/' ? (
+          <button onClick={handleStartQuiz} className="text-white">
+            start Quiz
+          </button>
+        ) : (
```
컴포넌트를 나누면 좋겠네요. 위에 함수들이 많은데 경로가 `/`일 때만 알아도 되는 내용과 `/`이 아닐 때 알아야 하는 내용이 섞여 있어서 읽기가 불편하고 나중에 `/`일 때 무언가 수정하러 코드를 보는 사람이 `/`이 아닐 때의 맥락까지 다 읽어야 해요.

new-deni commented on src/components/Header.tsx:
```diff
@@ -0,0 +1,100 @@
+          <>
+            <button
+              onClick={handlePrevious}
+              disabled={!currentQuizId || currentQuizId <= 1}
```
currentQuizId가 있다고 보장해주지 않으니 이런 코드가 계속 여기저기 나오고 있죠?

new-deni commented on src/components/Header.tsx:
```diff
@@ -0,0 +1,100 @@
+              <FaRegArrowAltCircleLeft
+                className="cursor-pointer text-customYellow"
+                size={24}
+              ></FaRegArrowAltCircleLeft>
```
self closing tag 사용해주세요

new-deni commented on src/components/Header.tsx:
```diff
@@ -0,0 +1,100 @@
+            ) : error ? (
+              <p className="text-red-500">{error}</p>
+            ) : (
+              <p className="text-white">{currentQuiz?.title}</p>
```
로딩도 에러도 아닌 상황이라서 이 코드까지 왔는데 nullish operator를 써야 하는군요? 개선할 점이 있다고 스스로 느끼실 수 있으면 좋습니다.

new-deni commented on src/components/Footer.tsx:
```diff
@@ -0,0 +1,7 @@
+  return <div>Footer</div>;
+};
+
+export default Footer;
```
default export와 named export 차이를 조사해보세요. 취향의 영역일 수 있지만 저는 named export을 사용합니다.

new-deni commented on tailwind.config.js:
```diff
@@ -2,7 +2,12 @@
-    extend: {},
+    extend: {
+      colors: {
+        customYellow: '#F7DF1E',
+        customGray: '#2F3035',
```
의미 있는 이름을 지어주세요. 기존 형식을 따라가는 게 좋습니다.

new-deni commented:
pr은 하나로 만들더라도 커밋은 나눠주세요.
```
라우트팅 설정
Header 컴포넌트 퀴즈 이동 버튼 기능 구현
라우터 설정 및 Home  → QuizPage 이동
컴포넌트 생성(Header, Footer, Layout, Quiz, QuizEditor, QuizVisualizaiotn, sidebar, Home, QuizPage)
임시데이터 생성 (data.ts)
```
위 내용들이 각각의 커밋이었어야 합니다.

nemo0824 committed:
chore:
react import 코드 삭제. (jsx 트랜스폼 도입) (footer, layout, quiz, quizEditor, quizVisualization, sidebar)
self closing tag 사용 (header)
fragment 태그 삭제  (app)

nemo0824 committed:
style:
커스텀 색상 이름  tailwind 형식에 맞게 변경

nemo0824 committed:
fix:
params를 사용하고 currentQuizId를 없앰으로서 상태관리 줄이기
FIRST_QUIZ_ID 로 1 이라는 매직넘버 정의 및 통일
loadingQuiz, fetchQuizData 비동기함수로 실제 통신시 동작할수있도록 변경
paseQuizId에 사용자가 예상치 못한 값 입력시 기본 url로 이동.

nemo0824 committed:
refactor:
default export 방식을 named export 방식으로 변경

nemo0824 committed:
refactor:
homeHeader 컴포넌트 분리

nemo0824 committed:
refactor:
상태관리 currentQuiz의 기본값 설정
기본값 보장으로 nullish operator 사용 지양

new-deni commented on src/components/Footer.tsx:
```diff
@@ -0,0 +1,5 @@
+
```
첫 줄은 왜 빈 줄로 남겨놓으셨나요..?

new-deni commented on src/components/Header.tsx:
```diff
@@ -0,0 +1,119 @@
+export const Header = () => {
+  const navigate = useNavigate();
+  const { quizId } = useParams();
+  const FIRST_QUIZ_ID = 1;
```
1. 컴포넌트 밖으로 빼는 게 더 자연스럽습니다.
2. 퀴즈 아이디는 원래 데이터베이스에 있는 값일 테고, 스트링 타입일 가능성도 높습니다. 결국 클라이언트에서 퀴즈아이디 기본값을 알 가능성은 낮아요. 설계 자체가 조금 바뀌면 좋겠네요.
위 내용이 와닿지 않는다면 일단 넘어가세요. 멘토링 세션 때 추가로 이야기 나누겠습니다.

new-deni commented on src/components/Header.tsx:
```diff
@@ -0,0 +1,119 @@
+  });
+
+
+  const fetchQuizData = async (id: number) => {
+    return new Promise((resolve, reject) => {
+      setTimeout(() => {
+        const quiz = quizData.find((q) => q.id == id);
+        if (quiz) {
+          setCurrentQuiz(quiz)
+          setError(null)
+          resolve(quiz)
+        } else {
+          setError('오류 퀴즈가 없어용');
+          reject(new Error('퀴즈를 찾을 수 없습니다.'));
+        }
+      }, 0)
+    })
+  };
```
`setCurrentQuiz`, `setError`와 같은 일을 `fetchQuizData`가 하는 게 맞을까요?
